Add std::variant backport as c10::variant

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26837 [BC-breaking] Use torch::variant for enums in C++ API
* **#26836 Add std::variant backport as c10::variant**

Differential Revision: [D17649064](https://our.internmc.facebook.com/intern/diff/D17649064)

Commit:
Add std::variant backport as torch::variant

ezyang commented:
I don't think variant.h should live in the torch. c10/util seems like a much more appropriate spot.

yf225 commented:
@ezyang I was a bit concerned about moving it to c10 because `variant` is not used in c10/ATen at the moment, but moving it seems to make code organization better.

Commit:
Update on "Add std::variant backport as torch::variant"

Add std::variant backport as torch::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

Commit:
Update on "Add std::variant backport as torch::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

ezyang commented on c10/util/variant.h:
```diff
@@ -0,0 +1,2824 @@
+    class destructor;
+
+#define MPARK_VARIANT_DESTRUCTOR(destructible_trait, definition, destroy) \
+  template <typename... Ts>                                
```
The way we have historically done these ports is to literally rename the header into our namespace, without leaving behind the old namespace. The reason for this is if libtorch is used in a library that itself is using mpark variant, if you put the implementation in the old mpark namespace, you will have two incompatible definitions of the same code, and compilation will fail.

yf225 replied:
Sounds great, I updated it accordingly.

ezyang reviewed (changes requested):
variant.h needs to be renamespaced directly into c10

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

ezyang commented on c10/util/variant.h:
```diff
@@ -2817,7 +2817,7 @@ namespace std {
 namespace c10 {
 
-using namespace mpark;
+using namespace c10;
```
Drop this now

ezyang reviewed (approved):
The tests are in the wrong place but I'll go ahead and approve this.

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

yf225 commented:
I moved the test to `aten/src/ATen/test/variant_test.cpp`.

kostmo commented:
This appears to have broken `caffe2_py2_clang3_8_ubuntu16_04_build` with the following error:
```
variant_test.cpp:15:23: error: default initialization of an object of const type 'const enumtype::Enum3' without a user-provided default constructor
```
Reverting.

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head

Commit:
Update on "Add std::variant backport as c10::variant"

Add std::variant backport as c10::variant

gh-metadata: pytorch pytorch 26836 gh/yf225/6/head
